src/main.rs

```diff
@@ -34,321 +33,323 @@ extern crate xi_core_lib;
 extern crate xi_rpc;
 #[macro_use]
 extern crate xi_win_shell;
+extern crate xi_win_ui;
 
-mod dialog;
 mod edit_view;
 mod linecache;
 mod menus;
 mod rpc;
 mod textline;
 mod xi_thread;
 
-use std::any::Any;
-use std::cell::{RefCell, RefMut};
-use std::rc::Rc;
 use std::sync::{Arc, Mutex};
-
-use winapi::shared::windef::*;
+use std::collections::HashMap;
 
 use serde_json::Value;
 
 use edit_view::EditView;
 use menus::MenuEntries;
 use rpc::{Core, Handler};
-use xi_win_shell::util::Error;
-use dialog::{get_open_file_dialog_path, get_save_file_dialog_path};
 use xi_thread::start_xi_thread;
 
-use xi_win_shell::paint::PaintCtx;
-use xi_win_shell::win_main::{self, RunLoopHandle};
-use xi_win_shell::window::{Cursor, IdleHandle, MouseEvent, WindowBuilder,
-    WindowHandle, WinHandler};
+use xi_win_shell::win_main::{self};
+use xi_win_shell::window::{Cursor, IdleHandle, WindowBuilder};
 
-struct MainWinState {
-    edit_view: EditView,
-}
+use xi_win_ui::{UiMain, UiState};
+use xi_win_ui::Id;
+use xi_win_ui::{FileDialogOptions, FileDialogType};
 
-impl MainWinState {
-    fn new() -> MainWinState {
-        MainWinState {
-            edit_view: EditView::new(),
-        }
-    }
+use edit_view::EditViewCommands;
 
-    fn render(&mut self, p: &mut PaintCtx) {
-        self.edit_view.render(p);
-    }
-}
+type ViewId = String;
 
-struct MainWinHandler {
-    win: Rc<MainWin>,
+#[derive(Clone)]
+struct ViewState {
+    id: Id,
+    filename: Option<String>,
+    handle: Arc<Mutex<IdleHandle>>,
```

Comment:
It's not necessary to `Arc` or `Mutex` this, as `IdleHandle` is `Clone`, and the methods on it are thread-safe (take `&self` rather than requiring `&mut self`).